chore: Add tests to ensure that build works and is functional

camilamacedo86 opened the pull request:
## Description
This PR introduces new tests aimed at building binaries for all supported platforms. The construction of these binaries utilizes make commands, ensuring their functionality. For this process, we've employed GitHub Actions and QEMU.

## Key Points:

- These additions appear sufficient in covering various scenarios like allowing us to get merged bumps without breaking changes without overthinking and manual tests.

mwasilew commented:
+1

vkhoroz reviewed (commented):
LGTM

Only a couple of minor comments.

vkhoroz commented on .github/workflows/binaries.yml:
```diff
@@ -0,0 +1,75 @@
+
+      - name: Install QEMU user static
+        if: matrix.qemu
+        run: sudo apt-get update && sudo apt-get install -y qemu-user-static
```
TBH I'm not a big fun of doing the QEMU thing ATM.
But, if the team agrees to this I'm fine.

camilamacedo86 commented on .github/workflows/binaries.yml:
```diff
@@ -0,0 +1,75 @@
+        uses: actions/upload-artifact@v2
+        with:
+          name: binaries
+          path: bin/
```
good point

camilamacedo86 replied:
done

camilamacedo86 commented on .github/workflows/binaries.yml:
```diff
@@ -0,0 +1,75 @@
+
+      - name: Verify binary
+        run: |
+          if ("${{ matrix.qemu }}" -eq "true") {
+            qemu-${{ matrix.qemu-platform }}-static ./bin/${{ matrix.binary }} logout
+          } else {
+            ./bin/${{ matrix.binary }} logout
+          }
```
I am fine with that. 
can be as well.

camilamacedo86 replied:
done

camilamacedo86 commented on .github/workflows/binaries.yml:
```diff
@@ -0,0 +1,75 @@
+          go-version: '~1.19'
+
+      - name: Build All Binaries
+        run: make
```
done

camilamacedo86 commented on .github/workflows/binaries.yml:
```diff
@@ -0,0 +1,75 @@
+          } else {
+            ./bin/${{ matrix.binary }} logout
+          }
+        shell: pwsh
```
done

camilamacedo86 commented:
Hi @vkhoroz ,

Thank you for taking the time to review this. I've addressed and incorporated all your suggestions. I believe it's now in a state ready for merging.

> TBH I'm not a big fan of doing the QEMU thing ATM. But, if the team agrees to this I'm fine.

As it stands, QEMU seems to be the most straightforward and cost-effective method for ARM testing without incurring any costs. Do you have any other alternatives in mind that can be conveniently integrated into GitHub actions, have widespread adoption, and are free?

Additionally, it's worth noting that we do recommend QEMU to our customers for testing. So, it aligns with our current recommendations.

vkhoroz commented on .github/workflows/binaries.yml:
```diff
@@ -0,0 +1,66 @@
+        run: chmod +x ./bin/${{ matrix.binary }}
+
+      - name: Verify binary
+        run: ./bin/${{ matrix.binary }} help
```
I planned to write a comment that QEMU part of the script was lost.

But, it somehow magically works without QEMU, right?
FYI in my local env the `fioctl-linux-arm64 help fails to load`, so I'm puzzled how this could work at all.

Is there any more magic to the `setup-qemu-action` in play here?

camilamacedo86 replied:
Hi @vkhoroz,

You raise a good point 🥇 . 
I removed the QEMU setup and did a test, and I observed that the arm64 failed:
[View the action run](https://github.com/camilamacedo86/fioctl/actions/runs/6386727467/job/17334072060)

![Screenshot 2023-10-03 at 01 00 04](https://github.com/foundriesio/fioctl/assets/7708031/74adfcbb-6d39-4606-ad84-66f18a9e116d)
![Screenshot 2023-10-03 at 00 59 52](https://github.com/foundriesio/fioctl/assets/7708031/b6f66b63-0f4e-44d0-818b-8f32c468f4e5)

Therefore, it seems that keeping only the following works **but it does not make sense for me either**:

```yaml
      - name: Set up QEMU
        if: matrix.qemu
        uses: docker/setup-qemu-action@v3
        with:
          platforms: ${{ matrix.qemu-platform }}
```

However, in my opinion, it is not transparent and reliable. Hence, I believe we should use a static setup so that we are making it very clear how things are working.

```yaml
      - name: Install QEMU user static
        if: matrix.qemu
        run: sudo apt-get update && sudo apt-get install -y qemu-user-static
        shell: bash

      - name: Verify binary
        run: |
          if [[ "${{ matrix.qemu }}" == "true" ]]; then
            qemu-${{ matrix.qemu-platform }}-static ./bin/${{ matrix.binary }} help
          else
            ./bin/${{ matrix.binary }} help
          fi
        shell: bash
```

Let me know your thoughts.
Can we move forward with this one now or would you like to propose any further change?

vkhoroz replied:
I like how the PR looks like now.
Good to know that action _really does the magic_; it makes me feel relieved.

IMHO using apt install directly is way more unreliable, at least it is not cross-platform.

So, let's merge the shorter version (before your latest commit below).

camilamacedo86 replied:
Done as suggested.

vkhoroz reviewed (approved):
LGTM
